### DISCUSSION TITLE
Question: how could I compile code with source generator

### DISCUSSION THREAD

**Discussion by WeihanLi** [dotnet/roslyn#67801] [Q&A] [answered]
For example, I want to compile the code like below

``` c#
Console.WriteLine($"Hello is {RegexHelper.IsLowercase("Hello")}");

public partial class RegexHelper
{
    // The Source Generator generates the code of the method at compile time
    [System.Text.RegularExpressions.GeneratedRegex("^[a-z]+$")]
    public static partial Regex LowercaseLettersRegex();

    public static bool IsLowercase(string value)
    {
        return LowercaseLettersRegex().IsMatch(value);
    }
}
```

When I try to compile the code, I got an error as follows(I had referenced the analyzer dlls, may in a wrong way):

```
CS8795-Error-(10,64): error CS8795: Partial method 'RegexHelper.LowercaseLettersRegex()' must have an implementation part because it has accessibility modifiers.
```

The code I used to compile the code: https://github.com/WeihanLi/SamplesInPractice/blob/ac8507c68390a6e7cc3daf1a5df35ee6b852ef86/RoslynSample/SourceGeneratorSample.cs

Are there some examples on how to compile with the `AnalyzerReference` or `SourceGenerator`?

**Comment by thomasclaudiushuber** [comment]
Works with .NET 7.0, fails with .NET 6.0, as the `System.Text.RegularExpressions.GeneratedRegex` was introduced with .NET 7.0 only. 

The error that comes with .NET 6.0 regarding the access modifier is because that attribute is missing in .NET 6.0.

<img width="861" alt="image" src="https://user-images.githubusercontent.com/7116546/231823927-06056035-f64f-4dca-91cf-bec6248ab199.png">

**Comment by WeihanLi** [comment]
@thomasclaudiushuber yeah, the regex source generator is introduced since .NET 7, it works for .NET 7 and above.
And my problem is I want to compile the code in runtime with Roslyn which failed to compile and I test with a .NET 8 console app

There's the code I used to compile the code

https://github.com/WeihanLi/SamplesInPractice/blob/ac8507c68390a6e7cc3daf1a5df35ee6b852ef86/RoslynSample/SourceGeneratorSample.cs

```c#
var code = @"
        using System;
        using System.Text.RegularExpressions;

        Console.WriteLine($""Hello is {RegexHelper.IsLowercase(""Hello"")}"");
        
        public partial class RegexHelper
        {
            [System.Text.RegularExpressions.GeneratedRegex(""^[a-z]+$"")]
            public static partial Regex LowercaseLettersRegex();

            public static bool IsLowercase(string value)
            {
                return LowercaseLettersRegex().IsMatch(value);
            }
        }";

var syntaxTree = CSharpSyntaxTree.ParseText(code);

var assemblyName = $"MyApp.{Guid.NewGuid()}";

// Add a reference to the System.Text.RegularExpressions assembly
var references = new MetadataReference[]
{
    MetadataReference.CreateFromFile(typeof(object).Assembly.Location),
    MetadataReference.CreateFromFile(typeof(Console).Assembly.Location),
    MetadataReference.CreateFromFile(typeof(System.Text.RegularExpressions.Regex).Assembly.Location),
};

// reference analyzer dll contains the source generator, harded for local testing purpose only
var generatorProjectPath = @"C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\8.0.0-preview.3.23174.8\analyzers\dotnet\cs\System.Text.Json.SourceGeneration.dll";
var projectReference = MetadataReference.CreateFromFile(generatorProjectPath);

var compilation = CSharpCompilation.Create(
    assemblyName,
    syntaxTrees: new[] { syntaxTree },
    references: references.Append(projectReference),
    options: new CSharpCompilationOptions(OutputKind.ConsoleApplication));

using var ms = new MemoryStream();
var result = compilation.Emit(ms);

if (result.Success)
{
    ms.Seek(0, System.IO.SeekOrigin.Begin);
    var assembly = Assembly.Load(ms.ToArray());
    assembly.EntryPoint?.Invoke(null, Array.Empty<object>());
}
else
{
    Console.WriteLine($"Compilation errors:");
    foreach (var diagnostic in result.Diagnostics)
    {
        Console.WriteLine(diagnostic.ToString());
    }
}
```

**Comment by CyrusNajmabadi** [comment]
> There's the code I used to compile the code

I don't see you using generators anywhere there?  You'll need to use [GeneratorDriver](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs) to actually run generators.

**Comment by thomasclaudiushuber** [comment, marked as answer]
Hi @WeihanLi ,

OK, got it. I looked at your project. You're creating the raw compilation object. On that one, the source generator did not run. That's the reason why you're getting the error.

You need to kick it off with a GeneratorDriver like I do here in this file: https://github.com/thomasclaudiushuber/mvvmgen/blob/main/src/MvvmGen.SourceGenerators.Tests/ViewModelGeneratorTests/Base/ViewModelGeneratorTestBase.cs

Then you'll get another Compilation object that includes the generated sources. 

It works like this for your project:

```
 var inputCompilation = CSharpCompilation.Create(
            assemblyName,
            syntaxTrees: new[] { syntaxTree },
            references: references,
            options: new CSharpCompilationOptions(OutputKind.ConsoleApplication));

var generatorAssemblyPath = @"C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\8.0.0-preview.3.23174.8\analyzers\dotnet\cs\System.Text.RegularExpressions.Generator.dll";
var generatorAssembly =  Assembly.LoadFile(generatorAssemblyPath);

var generator = generatorAssembly.GetType("System.Text.RegularExpressions.Generator.RegexGenerator");

var generatorInstance = generator.CreateInstance<IIncrementalGenerator>();

GeneratorDriver driver = CSharpGeneratorDriver.Create(generatorInstance);

driver = driver.RunGeneratorsAndUpdateCompilation(inputCompilation, out var outputCompilation, out var diagnostics);


using var ms = new MemoryStream();
var result = outputCompilation.Emit(ms);
```

You can close this issue, as it's not a .NET runtime issue.

I've sent you a PR with the working code in your repo: https://github.com/WeihanLi/SamplesInPractice/pull/1

Thank you for raising this issue, I'm sure it will help others when they read along here,
Thomas
